Read LHE reweighting automatically

ktht opened the pull request:
See: https://github.com/cms-nanoAOD/cmssw/issues/330
Creates new branches (`nLHEReweightingWeight` and `LHEReweightingWeight`). Tested on [1] and [2]. In both cases, the branches are created, and they're filled only if there's a weight group of the name `mg_reweighting`. Adding an illustration b/c the bot doesn't run on a single-top sample IIRC:

![Screenshot_20190503_045458](https://user-images.githubusercontent.com/6233872/57116973-4184fa80-6d61-11e9-8e32-64addc03798e.png)

[1] `/store/mc/RunIIFall17MiniAODv2/THQ_ctcvcp_4f_Hincl_13TeV_madgraph_pythia8/MINIAODSIM/PU2017_12Apr2018_94X_mc2017_realistic_v14-v1/00000/D400E164-D5F2-E811-9641-E0071B7B2320.root`
[2] `/store/mc/RunIIFall17MiniAODv2/ttHJetToNonbb_M125_TuneCP5_13TeV_amcatnloFXFX_madspin_pythia8/MINIAODSIM/PU2017_12Apr2018_new_pmx_94X_mc2017_realistic_v14-v1/100000/9A4C078D-C66A-E811-B217-0CC47A7C3434.root`

ktht committed:
Read LHE reweighting automatically

peruzzim commented:
@ktht would you have the chance to manually test it on samples for 2016 and 2018 as well, please?

ktht committed:
Accept any name for the reweighting weight

Previously only weights satisfying format 'rwgt_%d' were accepted, but in practice there can be more names for the reweighting weights (like 'mg_reweight_%d'). For this reason, it makes sense to accept any name for the weight (that's not an empty string).

ktht committed:
Remove useless regex from GenWeightsTableProducer

ktht commented:
Ok, tested in CMSSW release 10_2_10 with two threads and the results look good to me. I put the test results and script (`test_reweight.py`) to [1] and [2].

I found yet another error: `LHEEventProduct` name varies from sample to sample. Initially, I assumed that in all samples produced for 2016 era the name of `LHEEventProduct` is `source`, and in all 2017 & 2018 samples the expected name is `externalLHEProducer`. Currently the NanoAOD production assumes only `externalLHEProducer` as the name. Not only that, there seem to be some exceptions to the rule (see [3]). I worked around this problem in the test by querying `edmDumpEventContent`, then overwriting `genWeightsTable.lheInfo` and `lheInfoTable.lheInfo` with `--customise_commands` option in `cmsDriver.py`, but obviously this method doesn't scale w/ the official nor private Ntuple productions (because who knows to look up such things).

I'll see if I can figure out an elegant solution to this problem. The simplest way is to duplicate `lheInfo` to `lheInfo_2016`, but this doesn't scale in the long term if the name changes again.

[1] `/afs/cern.ch/user/k/kaehatah/public/LHE_fix.tar.gz`
[2] `/afs/cern.ch/user/k/kaehatah/public/LHE_fix (same as [1] but unpacked)`
[3]
```bash
edmDumpEventContent root://cms-xrd-global.cern.ch//store/mc/RunIISummer16MiniAODv2/THQ_ctcvcp_Hincl_13TeV-madgraph-pythia8_TuneCUETP8M1/MINIAODSIM/PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6-v1/80000/20C63FC3-30C6-E611-BE24-0CC47A537688.root | grep -i lhe
#LHEEventProduct                       "externalLHEProducer"       ""                "SIM"
```

ktht committed:
Support multiple names for LHEEventProduct

ktht commented:
Alright, I implemented the fix like so: https://github.com/HEP-KBFI/cmssw/compare/LHE_fix...HEP-KBFI:LHE_fix_2
Re-ran the tests (available in [1] and [2]) and everything looks good to me.

@peruzzim is it fine to include the fix to this PR or do you expect the fix to be implemented in a separate PR?

Btw, there seems to be yet another non-standard LHE header in dataset [3] (see [the attachment](https://github.com/cms-nanoAOD/cmssw/files/3146624/ggf_spin0_750_hh_2b2v.log)): there are 44 LHE scale weights (5 different scale choices per one pair of renomalization and factorization scales, 1 is not parsed correctly -> 45 - 1 = 44 scale weights).

[1] `/afs/cern.ch/user/k/kaehatah/public/LHE_fix_2.tar.gz`
[2] `/afs/cern.ch/user/k/kaehatah/public/LHE_fix_2` (same as [1] but unpacked)
[3] `/GluGluToRadionToHHTo2B2VTo2L2Nu_M-750_narrow_13TeV-madgraph_correctedcfg/RunIIFall17MiniAODv2-PU2017_12Apr2018_94X_mc2017_realistic_v14-v1/MINIAODSIM`

peruzzim commented:
> @peruzzim is it fine to include the fix to this PR or do you expect the fix to be implemented in a separate PR?

Yes please add that fix to this PR, after verifying whether it's really necessary to capture the `this` pointer in the lambda functions. Thanks
